code/API_definitions/qos-booking-and-assignment.yaml

```
@@ -0,0 +1,1756 @@
+      type: string
+      pattern: '^\+[1-9][0-9]{4,14}$'
+      example: "+123456789"
+
+    DeviceIpv4Addr:
+      type: object
+      description: |
+        The device should be identified by either the public (observed) IP address and port as seen by the application server, or the private (local) and any public (observed) IP addresses in use by the device (this information can be obtained by various means, for example from some DNS servers).
+
+        If the allocated and observed IP addresses are the same (i.e. NAT is not in use) then  the same address should be specified for both publicAddress and privateAddress.
+
+        If NAT64 is in use, the device should be identified by its publicAddress and publicPort, or separately by its allocated IPv6 address (field ipv6Address of the Device object)
+
+        In all cases, publicAddress must be specified, along with at least one of either privateAddress or publicPort, dependent upon which is known. In general, mobile devices cannot be identified by their public IPv4 address alone.
+      properties:
+        publicAddress:
+          $ref: "#/components/schemas/SingleIpv4Addr"
+        privateAddress:
+          $ref: "#/components/schemas/SingleIpv4Addr"
+        publicPort:
+          $ref: "#/components/schemas/Port"
+      anyOf:
+        - required: [publicAddress, privateAddress]
+        - required: [publicAddress, publicPort]
+      example:
+        {
+          "publicAddress": "203.0.113.0",
+          "publicPort": 59765
+        }
+
+    SingleIpv4Addr:
+      description: A single IPv4 address with no subnet mask
+      type: string
+      format: ipv4
+      example: "203.0.113.0"
+
+    DeviceIpv6Address:
+      description: |
+        The device should be identified by the observed IPv6 address, or by any single IPv6 address from within the subnet allocated to the device (e.g. adding ::0 to the /64 prefix).
+      type: string
+      format: ipv6
+      example: 2001:db8:85a3:8d3:1319:8a2e:370:7344
+
+    Status:
+      description: |
+        The current status of the requests in any of the endpoints in Booking or Assignment. The status can be one of the following:
+        * `PENDING` - The request is stil being processed. For booking, booking is requested and for assignements, the request is pending. 'statusInfo' may have additional information.
+        * `SUCCESSFUL` - The request is processed successfully. This may happen immediately.  The state will go to PENDING and then back to SUCCESSFUL in certain conditions.ƒ
+        * 'PARTIAL_SUCCESS' -  The request is processed with partial success - mostly applicable to device assignments.  'statusInfo' may have additional information. This enum is optional, and useful if the operator wants to communicate partical success.

```

```suggestion
        * `PARTIAL_SUCCESS` -  The request is processed with partial success - applicable to device assignments.  'statusInfo' may have additional information. This enum is optional, and useful if the operator wants to communicate partial success.
```
any case applicable outside device assignments? 

I have doubts around the expectations for the lifecycle of property `status`, which is returned with every BookingInfo and it should represent the status of the Booking. For example you assign several devices and some fail, then status=PARTIAL_SUCCESS, then you add one or many more and all succeed, is status still kept as PARTIAL_SUCCESS or reflects the result of the latest request only. And if a subsequent asssignment fails, is status  moved to FAILURE or still keeps the past PARTIAL_SUCCESS.